### DISCUSSION TITLE
Softbody explodes depending on node positions

### DISCUSSION THREAD

**Discussion by RauchMatthias** [bulletphysics/bullet3#3885] [General]
Hello,
there seems to be a problem with btDeformableNeoHookeanForce and btDeformableLinearElasticityForce. Depending on the position of the nodes OR the order of the nodes in the tetra, the softbody grows exponentially in Volume. I modified the "VolumetricDeformable.cpp" sample file to reproduce the issue using a softbody consisting of 4 nodes and a single tetra.

Replace the "create volumetric softbody"-part of the "initPhysics()" function with the following:

`
// create volumetric soft body
    {
       // btSoftBody* psb = btSoftBodyHelpers::CreateFromTetGenData(getDeformableDynamicsWorld()->getWorldInfo(),
       //                                                           TetraCube::getElements(),
       //                                                           0,
       //                                                           TetraCube::getNodes(),
       //                                                           false, true, true);
		//
        const int numNodes = 4;
		btAlignedObjectArray<btVector3> x;
		x.resize(numNodes);

        //explodes
		x[0] = btVector3(0, 0, 0);
		x[1] = btVector3(1, 0, 0);
		x[2] = btVector3(0, 0, 1);
		x[3] = btVector3(0, 1, 0);
		//works
		//x[0] = btVector3(0, 1, 0);
		//x[1] = btVector3(1, 0, 0);
		//x[2] = btVector3(0, 0, 1);
		//x[3] = btVector3(0, 0, 0);
		

		btSoftBody* psb = new btSoftBody(&getDeformableDynamicsWorld()->getWorldInfo(), numNodes, &x[0], 0);
		

		psb->appendTetra(0, 1, 2, 3); //explodes
		//psb->appendTetra(3, 1, 2, 0);  //works

		for (int i = 0; i < psb->m_tetras.size(); ++i)
		{
			psb->appendLink(psb->m_tetras[i].m_n[0], psb->m_tetras[i].m_n[1], 0, true);
			psb->appendLink(psb->m_tetras[i].m_n[1], psb->m_tetras[i].m_n[2], 0, true);
			psb->appendLink(psb->m_tetras[i].m_n[2], psb->m_tetras[i].m_n[0], 0, true);
			psb->appendLink(psb->m_tetras[i].m_n[0], psb->m_tetras[i].m_n[3], 0, true);
			psb->appendLink(psb->m_tetras[i].m_n[1], psb->m_tetras[i].m_n[3], 0, true);
			psb->appendLink(psb->m_tetras[i].m_n[2], psb->m_tetras[i].m_n[3], 0, true);
		}

		psb->initializeDmInverse();
		psb->m_tetraScratches.resize(psb->m_tetras.size());
		psb->m_tetraScratchesTn.resize(psb->m_tetras.size());

        getDeformableDynamicsWorld()->addSoftBody(psb);
        psb->scale(btVector3(2, 2, 2));
        psb->translate(btVector3(0, 15, 0));
        psb->getCollisionShape()->setMargin(0.1);
        psb->setTotalMass(0.5);
        psb->m_cfg.kKHR = 1; // collision hardness with kinematic objects
        psb->m_cfg.kCHR = 1; // collision hardness with rigid body
		psb->m_cfg.kDF = 2;
        psb->m_cfg.collisions = btSoftBody::fCollision::SDF_RD;
        psb->m_cfg.collisions |= btSoftBody::fCollision::SDF_RDN;
		psb->m_sleepingThreshold = 0;
        btSoftBodyHelpers::generateBoundaryFaces(psb);
        btDeformableGravityForce* gravity_force =  new btDeformableGravityForce(gravity);
        getDeformableDynamicsWorld()->addForce(psb, gravity_force);
        m_forces.push_back(gravity_force);
        

		//btDeformableNeoHookeanForce* neohookean = new btDeformableNeoHookeanForce(20, 80, .01);
        btDeformableLinearElasticityForce* linearElasticity = new btDeformableLinearElasticityForce(100,100,0.01);
		m_linearElasticity = linearElasticity;
		getDeformableDynamicsWorld()->addForce(psb, linearElasticity);
        m_forces.push_back(linearElasticity);
    }`

Use the following lines to switch between working and exploding behavior:
psb->appendTetra(0, 1, 2, 3); //explodes
//psb->appendTetra(3, 1, 2, 0);  //works
Alternatively switch the position of node 0 and 3 before the softbody creation.

With btDeformableNeoHookeanForce it explodes in the air,  with btDeformableLinearElasticityForce it explodes when it hits the ground. Either way, I suspect forces being applied in the opposite direction.
btDeformableMassSpringForce does not have this problem.

**Comment by erwincoumans** [comment]
Unfortunately, the deformable code is an unsupported internship project. Not more we can do then just turn it into a discussion.
If you, or someone else, has time/resources to fix it, please let us know.

**Comment by RauchMatthias** [comment]
Ok, I found the reason, in btSoftbody::initializeDmInverse() the formula for calculating the tetra volume is missing the fabs(). The variable is called m_element_measure. 
But I think that this is a design choice. Apparently there is a convention to order vertices following the right-hand rule, so that the determinant is always positive. The cube and bunny meshes follow this convention, I looked at it. 
So, programmatically building a softbody works by calculating the determinant for each desired tetra, and if it is negative, switching any 2 of the nodes to make it positive. Then use btSoftbody::appendTetra.
For my application this is ok.
Therefore, if this is intended, there is no issue except maybe for the missing documentation. But at least it made me learn something.
